data_structures/doubly_linked_list.ts

```diff
@@ -0,0 +1,299 @@
+        } else {
+            this.head.next!.prev = null;
+        }
+
+        this.head = this.head.next;
+        this.length--;
+
+        return removedNode;
+    }
+
+    /**
+     * Inserts a node at the tail of the list.
+     *
+     * @param value The value of the node being inserted.
+     */
+    append(value: T): void {
+        const newNode = new DoublyLinkedListNode(value);
+
+        if (!this.head) {
+            this.head = newNode;
+        } else {
+            this.tail!.next = newNode;
+            newNode.prev = this.tail;
+        }
+
+        this.tail = newNode;
+        this.length++;
+    }
+
+    /**
+     * Removes a node from the tail of the list.
+     *
+     * @returns The node that was removed.
+     */
+    removeTail(): DoublyLinkedListNode<T> | null {
+        if (!this.head) {
+            return null;
+        }
+
+        const removedNode = this.tail;
+
+        if (this.head === this.tail) {
+            this.head = null;
+        } else {
+            this.tail!.prev!.next = null;
+        }
+
+        this.tail = this.tail!.prev;
+        this.length--;
+
+        return removedNode;
+    }
+
+    /**
+     * Inserts a node at a specific index.
+     *
+     * @param index The index where the node will be inserted.
+     * @param value The value of the node being inserted.
+     * @throws Index out of bounds if the index is not valid.
+     */
+    insertAt(index: number, value: T): void {
+        if (index < 0 || index > this.length) {
+            throw new Error("Index out of bounds");
+        }
+
+        if (index === 0) {
+            this.push(value);
+            return;
+        }
+
+        if (index === this.length) {
+            this.append(value);
+            return;
+        }
+
+        const newNode = new DoublyLinkedListNode(value);
+        const prevNode = this.get(index - 1)!;
+        const nextNode = prevNode.next;
+
+        prevNode.next = newNode;
+        newNode.prev = prevNode;
+        newNode.next = nextNode;
+        nextNode!.prev = newNode;
+
+        this.length++;
+    }
+
+    /**
+     * Removes a node at a specific index.
+     *
+     * @param index The index of the node to be removed.
+     * @returns The node that was removed.
+     * @throws Index out of bounds if the index is not valid.
+     */
+    removeAt(index: number): DoublyLinkedListNode<T> | null {
+        if (index < 0 || index >= this.length) {
+            throw new Error("Index out of bounds");
+        }
+
+        if (index === 0) {
+            return this.pop();
+        }
+
+        if (index === this.length - 1) {
+            return this.removeTail();
+        }
+
+        const removedNode = this.get(index)!;
+        removedNode.prev!.next = removedNode.next;
+        removedNode.next!.prev = removedNode.prev;
+
+        this.length--;
+
+        return removedNode;
+    }
+
+    /**
+     * Reverses the list.
+     *
+     * @returns The reversed list.
```

Comment:
It doesn't, the list is reversed in-place.

Reply:
I don't understand... What is the difference?

Reply:
> I don't understand... What is the difference?

The function doesn't return a new, reversed list. It doesn't even return itself. It just reverses itself in-place, returning nothing. The function is a `void`.